### ISSUE TITLE
Mig Alley Windows 10 Error

### ISSUE THREAD

**Issue by TheSilkMan** [narzoul/DDrawCompat#116]
Hi Narzoul, it appears the v0.3.1 dll breaks Mig Alley.  The dll from v0.2.1 still works. So, somewhere between the two versions a change has caused this.
The following error occurs: "The procedure entry point SymAddrIncludeInlineTrace could not be located in the dynamic link library C:\Windows\System32\dbeng.dll."

The log file reads:
4300 23:36:49.449 Process path: C:\migalley\MigAlley.exe
4300 23:36:49.449 Loading DDrawCompat from C:\migalley\DDRAW.dll
4300 23:36:49.451 DDrawCompat loaded successfully
4300 23:36:50.136 Installing DirectDraw hooks
4300 23:36:51.397 Installing GDI hooks
4300 23:36:51.408 Installing registry hooks
4300 23:36:51.408 Finished installing hooks
4300 23:40:22.485 GDI used DC cache size: 1

**Comment by narzoul** [comment]
If there's a dbghelp.dll file in your install directory, rename it or remove it. Some games ship with an old version of the file, but Windows has a much more recent version already, and the old version is incompatible with the newer dbgeng.dll in Windows. I will add a workaround to the next release that may help with this (force load the Windows version of dbghelp if possible).

**Comment by TheSilkMan** [comment]
Thanks. I did see that file there and will try removing it from the game
directory.  Great job removing that black box off of the screen.  How did
you find and accomplish that?  One other thing that happens is sometime
when exiting 3d from a mission, the game hangs. When I alt tab, I see the
game window with the black box, but it fails to exit to the 2D map. Have to
use taskmaster to force it to close.

Finally, have you found a way to fix the patchy holes in the cockpit
textures of the F-86?

**Comment by TheSilkMan** [comment]
Hi Narzoul, I removed the dgbhelp.dll file. The error no longer appears at
launch of the game.  However, with the new ddraw.dll the game starts to
launch, then goes directly back to desktop.  No events are recorded in the
log file.  The old ddraw.dll still works though.

**Comment by narzoul** [comment]
I haven't done anything about a "black box" because it was never there with DDrawCompat. Isn't it just an issue with other wrappers?
I still don't have an NVIDIA test system, so I can't do much about the cockpit issues. It doesn't happen on AMD or Intel. I asked for debug logs from an NVIDIA system a long time ago on Vogons, but never got them. I'm not sure it would be enough to figure out the problem anyway.

As for v0.3.1, do you mean there is no DDrawCompat-MigAlley.log file at all? Note that the log file name changed since v0.2.1.

**Comment by TheSilkMan** [comment]
Hi Narzoul, yes the black box occurs when you run the game on Windows 10
without any wrapper or with wrappers other DDrawCompat.  With DDrawCompat,
the box (which is designed to house the ui and swap z-order with 3d window
when in 3d)still occurs as with other dual windows games, but correctly
resizes and gets placed in correct Z-Order beneath the 3D window.

Yes, I am running an NVidia 3070 card. The missing cockpit texture pieces
also occurred in my older 980ti card.   If you tell me how to generate the
NVidia debug file(and where it saves to), I'll be sure to share it with you.

I double checked  the log file. It is being generated under the name
DDrawCompat-MigAlley.log in v0.3.1(contents in the attached). In this
version, the game never fully loads to the main UI. It immediately drops to
desktop. It is 100% repeatable and as such, the game is unplayable.

In v0.2.1, when playing a campaign, after flying a mission, exiting and the
game attempts to return you to the 2D map screen, the game hangs and never
loads the 2d map. This only happens with the  DDrawCompat wrapper. It is
100% repeatable.  It also happens periodically when exiting 3D and
returning to 2D map in the HotShot quick mission. I have not seen this with
other wrappers.

**Comment by narzoul** [comment]
I don't see any attachments. Maybe it doesn't work if you send it as an e-mail attachment. Try attaching it via GitHub.

v0.2.1 is more than 5 years old. The hanging issue has probably been fixed already. At least I haven't encountered it while testing this game recently.

**Comment by TheSilkMan** [comment]
Hi Narzoul,  thanks for the reply. Sounds good, I am hoping to get v0.3.1 working to see your progress.  Attached is a screen shot of the log file.
![20211003_132159](https://user-images.githubusercontent.com/49526415/135786385-041f6bf8-f317-4f67-b80d-8f91b6f8d438.jpg)

**Comment by narzoul** [comment]
Nothing unusual there so far. Please try collecting the logs using the ddraw.dll from the ReleaseWithDebugLogs build found in the debug release:
https://github.com/narzoul/DDrawCompat/releases/download/v0.3.1/DDrawCompat-v0.3.1-debug.zip

Also, if it's a crash, you might be able to get a crash dump following the instructions here:
https://github.com/narzoul/DDrawCompat/issues/50#issuecomment-520162844

Please attached both (preferably in a zip archive) if possible.

On a side note, why is your executable called MigAlley.exe instead of just Mig.exe? What patches have you installed?

Although I doubt it's related, Win98 compatibility mode should not be needed. You can use Windows 8 compatibility mode if the thick window borders in the game UI bother you, otherwise no compatibility mode should be set.

**Comment by TheSilkMan** [comment]
Hi Narzoul, I will give this a shot tonight.  Also, how do I obtain the
Nvidia specific report that you need?

In terms of the file and directory naming convention. I prefer to call
things what they are(hence MigAlley). The sim is about much more than just
the Mig. ;)

**Comment by narzoul** [comment]
It would need the logs from the ReleaseWithDebugLogs build also, but that comes only after getting v0.3.1. working. Just run it with that build in a quick mission for a few seconds and then exit and collect the logs.

**Comment by TheSilkMan** [comment]
Hi Narzoul, since it crashes immediately upon launch, I never get to go into quick mission wth v0.3.1.  However, I was able to grab todays crash log and the dump file(see attached
[Silk's Mig CrashLog.20904.zip](https://github.com/narzoul/DDrawCompat/files/7290080/Silk.s.Mig.CrashLog.20904.zip)
).

**Comment by TheSilkMan** [comment]
Hi Narzoul, that worked!  It now loads correctly to the menu and no CTD.  Another good thing is that after exiting campaign missions no longer cause as freeze/lockup as in v0.2.1.  Even better, frame rates are very high both in campaign and quick mission.  Another good thing is that the load screen graphic is much bigger and fills most of the screen. Before it did not enlarge and just stayed very small.
Log file:
[DDrawCompat-MigAlley.log](https://github.com/narzoul/DDrawCompat/files/7298025/DDrawCompat-MigAlley.log)

There is however one show stopping anomaly.  It now takes up to 7 minutes from launch of either a quick mission or campaign mission to load the 3d world. Perhaps, you included the the debug version of the dll?  Running in debug normally causes that type of delay.  

The other thing I noticed is that on load with this version, it causes that 2nd window for the UI to not follow Z-order and be hidden during the loading of the 3d world(visible with the load screen). This did not happen in v0.2.1. It resolves itself after load.  I have a 3 monitor setup so perhaps this is why I see it(on the 2nd monitor).This did not happen in v0.2.1. It resolves itself after load.  There is a little lag or menu trailing when you grab a menu and drag it to relocate it to another part of the desktop.  No thick borders though.

**Comment by TheSilkMan** [comment]
After that, just need to resolve the Nvidia missing cockpit texture issue.

One feature that would be nice is to be able to fill the three monitors with the full world, (i.e., maintain the aspect ratio without getting the distortion that can come with just stretching of the view).

**Comment by TheSilkMan** [comment]
Hi Narzoul, ok I was able to try it tonight.  Load times are the same(or longer). However, this version also causes the load status bar to not appear.  

Just for comparison, I loaded up the v0.2.1 again.  I noticed that the 0.2.1 actually only uses the center monitor when in 3d loading and 3d world(left and right monitors go to monitors default screen[aka blue screen on some machines], hence why I didn't see the overlapping box.  It appears in the far left monitor with the DDrawCompat dll.  With the default code in Windows 10, it just stays and overlaps smack over the 3D world.  Also, the Hotshot mission loads slowly like v0.3.1 and v0.3.1+, but loads much faster in Campaign. In  v0.2.1 framerates are subpar when compared to v0.3.1+.

Yes you are correct, I have a triple monitor setup. It is running in Nvidia surround mode which causes windows to see all three monitors as one giant 11520x2160 monitor(i.e., 3 screens each have 3840x2160). The game supports 3 monitors. However, 3d and driver have to be optimized to fully take advantage of it.  For now, a command line switch forcing it to load in the 11520x2160 can work.  I was able to exploit this feature in DxWnd. The drawback was the somewhat stretched aspect angle.  However, the game can be zoomed to somewhat correct for this using the built in wide screen zoom.  Its not perfect but it will fill the three monitors. Right now, 2d uses all three screens, 3d uses center only.

More on the Overlapping Window

The code pushes the 2d swapped (UI Window) off out of view(upper leftmost corner of a virtual left screen) during 3D(win95, XP,Win7).  Back then they assumed that people would not see it(out of sight out of mind). As such, it's not just a Z-Order issue as I thought.
 It would have been much better if it was just minimized when in 3d and maximized when in the UI.

**Comment by narzoul** [comment]
> Load times are the same(or longer)

Please try it with all CPU cores enabled (from Task Manager, after launching the game).

> With the default code in Windows 10, it just stays and overlaps smack over the 3D world.

The default mode for DirectDraw (since Vista probably) is an emulated borderless windowed mode (the application thinks it's running in exclusive full-screen mode, but it's actually windowed). So it might interact with window placement differently. Try it with the "DXPrimaryEmulation -DisableMaxWindowedMode" shim from ACT and see if there's a difference when it runs in real exclusive mode (like it does with DDrawCompat).

> The game supports 3 monitors. However, 3d and driver have to be optimized to fully take advantage of it.

I don't understand what you mean by this. But if it only works via Surround, then I'm pretty sure the game itself only sees/uses one "virtual monitor" at a time, the same as if it was running on a single monitor. If it works without Surround too, then let me know how to set it up, I'll try to test it with 2 monitors.

**Comment by narzoul** [comment]
There will be a config option in the next release for CPU affinity (among other things).
For now, you can use this test release, where I removed the single core affinity setting (in addition to the previous changes):
[ddraw.zip](https://github.com/narzoul/DDrawCompat/files/7332708/ddraw.zip) ([diff.txt](https://github.com/narzoul/DDrawCompat/files/7332711/diff.txt) compared to v0.3.1)

Since this should work well enough now, I've included a ReleaseWithDebugLogs build also to check out the cockpit transparency issue. Just reproduce it on screen in some quick mission for a few seconds, exit the game and upload the logs (compressed).

**Comment by TheSilkMan** [comment]
Hi Narzoul,  in both debug and non-debug versions, it loads in single core affinity mode. I verified this by checking the affinity of each. After the debug version finally loads and I exit, the log file is 89,155 kb and cannot be attached here. 

The non debug versions loads fast as usual when affinity is set to all.  When I set the affinity to all and load in the debug version(figured I'd give it a shot), it minimizes to the taskbar and no way to  maximize the game to full screen.

**Comment by TheSilkMan** [comment]
Yes, it was in compatibility mode. Unchecked that and it solved the problem.  The affinity fix now works. Below is the log from debug version. I could not do much looking around with the views since the file gets really large. I basically entered 3D, then exited.  The main cockpit texture in the F-86 is patchy with segments missing so it should show up in the log.

[DDrawCompat-MigAlley.log](https://github.com/narzoul/DDrawCompat/files/7356607/DDrawCompat-MigAlley.log)

**Comment by TheSilkMan** [comment]
"NVIDIA seems to default to black pixels as transparent." 
I wonder why they chose to do this.  Ok, I tried both and so far, no difference
in the the transparent pixels.  I also tried turning transparency on and off and the various filter options in the game but that doesn't seem to help.

Changing from primary(NVidia) driver to software driver stops the transparency issue, but (of course) you lose all Directx 3d effects.

**Comment by narzoul** [comment]
Strange that neither fix has any effect at all. I can reproduce the problem on my AMD GPU by forcing black to be the transparent color, instead of disabling color key transparency.

Well, that gives me another idea. Instead of trying to disable the transparency, let's see if we can change the transparent color to white instead:
[ddraw.zip](https://github.com/narzoul/DDrawCompat/files/7360436/ddraw.zip) ([diff.txt](https://github.com/narzoul/DDrawCompat/files/7360438/diff.txt) compared  to v0.3.1)

On my system, this makes only a few pixels (small rectangles) transparent on the top light-gray part of the cockpit. I hope it will have a similar effect for you, meanwhile fixing the black parts.

**Comment by TheSilkMan** [comment]
Hi Narzoul, you are definitely on to something.  This eliminates the missing black textures.  However, I do see the missing grey pieces on the range meter, the star on the wings, and the text in the exhaust temperature guage reverts to black(instead of white).

![1](https://user-images.githubusercontent.com/49526415/137638750-4ca83f54-db69-4b6e-a540-d774ca4b30b6.jpg)
![4](https://user-images.githubusercontent.com/49526415/137638751-55e51424-e256-42ea-be0e-2a54270deaab.jpg)
![3](https://user-images.githubusercontent.com/49526415/137638752-e8e28e2d-e73f-4d54-9bca-7a490ecc8c2c.jpg)
![2](https://user-images.githubusercontent.com/49526415/137638753-6f75cdbc-8ee8-4c25-b4c4-706e02f55343.jpg)

**Comment by narzoul** [comment]
Ok. Let's have one more test for today please, just to confirm my theory. So it seems the "disabling" color key transparency for textures means setting black as the color key for NVIDIA.

Here's a modified version of the previous patch:
[ddraw.zip](https://github.com/narzoul/DDrawCompat/files/7360605/ddraw.zip) ([diff.txt](https://github.com/narzoul/DDrawCompat/files/7360606/diff.txt) compared to v0.3.1)

After setting the color key to white, it immediately sets is back to disabled. Now it renders perfectly on my system, but I assume it's going to just revert back to having black as the transparent color on yours. If that's the case, I'm gonna have to come up with some creative workaround to fix this driver bug efficiently...

**Comment by TheSilkMan** [comment]
Hi Narzoul, it's pretty much the same on my end:


![5](https://user-images.githubusercontent.com/49526415/137644271-2e9184b5-05f9-46a6-a823-f6e9a2a68f10.jpg)
![6](https://user-images.githubusercontent.com/49526415/137644405-138da90c-6232-4e44-99d8-20e748ea38a9.jpg)

**Comment by narzoul** [comment]
The NVIDIA texture color key issues should be solved in v0.4.0.

**Comment by TheSilkMan** [comment]
Thanks a lot Narzoul. I will check it out.

**Comment by TheSilkMan** [comment]
Hi Narzoul, what other changes or features does 4.0 bring vs 3.1? For
instance, does it still feature the CPU Affinity fix?

**Comment by narzoul** [comment]
There is a setting for it. Read the [release notes](https://github.com/narzoul/DDrawCompat/releases/tag/v0.4.0).

**Comment by TheSilkMan** [comment]
Hi Narzoul, I tried v4 with Mig Alley. Unfortunately it CTDs after double
clicking the icon. Log says files are missing

**Comment by narzoul** [comment]
From reading back the discussion here, the game crashed for you with v0.3.1 also without some custom modifications, due to too many supported display modes. You can probably do those modifications now via config files. Check the [Configuration](https://github.com/narzoul/DDrawCompat/wiki/Configuration) wiki for details.

Based on the previous discussion, I'd suggest the following settings:
CpuAffinity=all
SupportedResolutions=640x480, 800x600, 1024x768
ThreadPriorityBoost=on

Just put them in a DDrawCompat.ini file in your game directory.

**Comment by thomasplummer55** [comment]
Hi Narzoul,

 I have followed this discussion with interest since I have been trying to run Mig Alley on Win10. I have now had good success with one of your ddraw.dll files and a standard BDG install of Mig Alley. Framerate is 60 everywhere and no CTDs or artifacts.
The specific ddraw.dll is the one referenced in this discussion with this text:

"Anyway, to verify this theory, I removed the enumeration of 8 and 32 bit display modes from v0.3.1 in this test version: ddraw.zip (diff.txt compared to v0.3.1)"

These show my system and graphics driver:


![mig_1](https://user-images.githubusercontent.com/21131589/226671009-677304dc-573c-48fb-8529-d6d449c3dd8d.png)
![Mig_2](https://user-images.githubusercontent.com/21131589/226671013-8631c937-6b3c-4f4b-9a39-5f8b8b42195d.png)
![mig_3](https://user-images.githubusercontent.com/21131589/226671016-759f006b-1c1e-436b-8f0a-f142141095fa.png)
![Mig_4](https://user-images.githubusercontent.com/21131589/226671042-29dfe1db-38c3-473b-9982-e543b71c69bf.png)


I dragged and dropped the pictures here but I just see encoded text, is there another way??

Thanks for your work on Mig Alley, I can say that it runs better than Win98 or WinXP which are 50 fps.
I can now fly on Win10 without the need to use my WinXP laptop.

Well done Narzoul,
Thomas
